course7/homeworks/patrickh/homework3/calculator.cpp

```diff
@@ -0,0 +1,73 @@
+#include <stdexcept>
+#include <iostream>
+#include "calculator.h"
+
+void Calculator::parse(const std::string& input)
+{
+    if (input == "+")
+        add();
+    else if (input == "-")
+        subtract();
+    else if (input == "*")
+        multiply();
+    else if (input == "/")
+        divide();
+    else
+        mStack.emplace(input);
+}
+
+void Calculator::add()
+{
+    if(mStack.size() < 2)
+        throw std::runtime_error("Not enough values to be added");
+    Complex c1 = mStack.top();
+    mStack.pop();
+    Complex c2 = mStack.top();
+    mStack.pop();
+    mStack.emplace(c1 + c2);
+}
+
+void Calculator::subtract()
+{
+    if(mStack.size() < 2)
+        throw std::runtime_error("Not enough values to be subtracted");
+    Complex c1 = mStack.top();
+    mStack.pop();
+    Complex c2 = mStack.top();
+    mStack.pop();
+    mStack.emplace(c1 - c2);
+}
+
+void Calculator::multiply()
+{
+    if(mStack.size() < 2)
+        throw std::runtime_error("Not enough values to be multiplied");
+    Complex c1 = mStack.top();
+    mStack.pop();
+    Complex c2 = mStack.top();
+    mStack.pop();
+    mStack.emplace(c1 * c2);
+}
+
+void Calculator::divide()
+{
+    if(mStack.size() < 2)
+        throw std::runtime_error("Not enough values to be divided");
+    Complex c1 = mStack.top();
+    mStack.pop();
+    Complex c2 = mStack.top();
+    mStack.pop();
+    mStack.emplace(c1 / c2);
+}
```

Comment:
This code here is correct but very verbose. Please, have a look at Patrick K's solution, he managed to squeeze it all in some 8 lines of code.